### PR TITLE
[nova] Topology aware scheduling and spreading + AZ update

### PR DESCRIPTION
**Pull request by fwiesel**
Use topologySpreadConstraints to spread the api pods over the zones in a best effort manner to improve the situation in the situation of a zone is down (kubernetes may not rebalance, or cannot due to inbalance)

Use topology aware routing service-annotation to make use of the distributed placement and prefer communicating to nodes in the same AZ

In regions with a single AZ, the annotations are not set.

### CODE REVIEW

File: openstack/nova/templates/console-shellinabox-deployment.yaml [hunk 1]
```diff
@@ -25,6 +25,7 @@ spec:
       labels:
         name: nova-console-cell1-shellinabox
 {{ tuple . "nova" "console-cell1-shellinabox" | include "helm-toolkit.snippets.kubernetes_metadata_labels" | indent 8 }}
```
Comment (joker-at-work) [inline openstack/nova/templates/console-shellinabox-deployment.yaml:27]: Optional: Move this to `nindent`, too?

File: openstack/nova/templates/console-shellinabox-deployment.yaml [hunk 2]
```diff
@@ -37,7 +38,8 @@ spec:
         runAsGroup: 33
         fsGroup: 33
 {{ tuple . "nova" "console-cell1-shellinabox" | include "kubernetes_pod_anti_affinity" | indent 6 }}
```
Comment (joker-at-work) [inline openstack/nova/templates/console-shellinabox-deployment.yaml:40]: Optional: Move this to `nindent`, too?

File: openstack/nova/templates/scheduler-deployment.yaml [hunk 3]
```diff
@@ -35,8 +35,8 @@ spec:
         configmap-etc-hash: {{ include (print $.Template.BasePath "/etc-configmap.yaml") . | sha256sum }}
         configmap-scheduler-etc-hash: {{ include (print $.Template.BasePath "/scheduler-etc-configmap.yaml") . | sha256sum }}
     spec:
-{{ tuple . "nova" "scheduler" | include "kubernetes_pod_anti_affinity" | indent 6 }}
-{{ include "utils.proxysql.pod_settings" . | indent 6 }}
+      {{- tuple . "nova" "scheduler" | include "kubernetes_pod_anti_affinity" | nindent 6 }}
+      {{- include "utils.proxysql.pod_settings" . | nindent 6 }}
```
Comment (joker-at-work) [inline openstack/nova/templates/scheduler-deployment.yaml:39]: Why does `nova-scheduler` not need `utils.topology.constraints`?

File: openstack/nova/templates/console-shellinabox-deployment.yaml [hunk 4]
```diff
@@ -37,7 +38,8 @@ spec:
...
-{{ include "utils.proxysql.pod_settings" . | indent 6 }}
+      {{- include "utils.proxysql.pod_settings" . | nindent 6 }}
+      {{- tuple . (dict "name" "nova-console-shellinabox") | include "utils.topology.constraints" | indent 6 }}
```
Comment (joker-at-work) [inline openstack/nova/templates/console-shellinabox-deployment.yaml:42]: I think we need to rename this to `nova-console-cell1-shellinabox`.

File: openstack/nova/templates/cell2-conductor-deployment.yaml [hunk 5]
```diff
@@ -28,15 +28,17 @@ spec:
         alert-tier: os
         alert-service: nova
 {{ tuple . "nova" "conductor" | include "helm-toolkit.snippets.kubernetes_metadata_labels" | indent 8 }}
```
Comment (joker-at-work) [inline openstack/nova/templates/cell2-conductor-deployment.yaml:30]: Optional: Move to `nindent`?

File: openstack/nova/templates/cell2-conductor-deployment.yaml [hunk 6]
```diff
@@ -28,15 +28,17 @@ spec:
...
         {{- if or .Values.cell2.conductor.config_file.DEFAULT.statsd_enabled .Values.proxysql.mode }}
         prometheus.io/scrape: "true"
         prometheus.io/targets: {{ required ".Values.alerts.prometheus missing" .Values.alerts.prometheus | quote }}
         {{- end }}
         configmap-etc-hash: {{ include (print $.Template.BasePath "/etc-configmap.yaml") . | sha256sum }}
     spec:
-{{ tuple . "nova" "conductor" | include "kubernetes_pod_anti_affinity" | indent 6 }}
-{{ include "utils.proxysql.pod_settings" . | indent 6 }}
+      {{ tuple . "nova" "conductor" | include "kubernetes_pod_anti_affinity" | nindent 6 }}
+      {{ include "utils.proxysql.pod_settings" . | nindent 6 }}
```
Comment (joker-at-work) [inline openstack/nova/templates/cell2-conductor-deployment.yaml:40]: Shouldn't they both start with `{{-`? Doesn't break anything, I think, just adds an additional newline.

**Review by joker-at-work — CHANGES_REQUESTED** [review]

File: openstack/nova/templates/console-shellinabox-deployment.yaml (hunk 1 above)
Comment (fwiesel) [reply openstack/nova/templates/console-shellinabox-deployment.yaml:27]: Done

File: openstack/nova/templates/console-shellinabox-deployment.yaml (hunk 2 above)
Comment (fwiesel) [reply openstack/nova/templates/console-shellinabox-deployment.yaml:40]: Done

File: openstack/nova/templates/scheduler-deployment.yaml (hunk 3 above)
Comment (fwiesel) [reply openstack/nova/templates/scheduler-deployment.yaml:39]: We only run one instance of it, and if we run more than one instance, we might get into trouble. AFAIK.

File: openstack/nova/templates/console-shellinabox-deployment.yaml (hunk 4 above)
Comment (fwiesel) [reply openstack/nova/templates/console-shellinabox-deployment.yaml:42]: done

File: openstack/nova/templates/cell2-conductor-deployment.yaml (hunk 6 above)
Comment (fwiesel) [reply openstack/nova/templates/cell2-conductor-deployment.yaml:40]: Changed it

**Comment by fwiesel** [comment]
The last update pulls in a newer `utils` which uses the non-deprecated topology key.

File: openstack/nova/templates/scheduler-deployment.yaml (hunk 3 above)
Comment (joker-at-work) [reply openstack/nova/templates/scheduler-deployment.yaml:39]: hm ... correct, we currently do, but we also define `kubernetes_pod_anti_affinity` anyways, so I would have expected differently. anyways, no need to change currently, I guess. thanks.

**Review by joker-at-work — APPROVED** [review]

**Pull request merged**